File: src/test/java/org/scijava/ops/core/builder/OpBuilderTest.java
/*
 * #%L
 * SciJava Operations: a framework for reusable algorithms.
 * %%
 * Copyright (C) 2016 - 2019 SciJava Ops developers.
 * %%
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 * #L%
 */

package org.scijava.ops.core.builder;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import java.util.Arrays;
import java.util.function.BiFunction;
import java.util.function.Function;

import org.junit.Test;
import org.scijava.ops.AbstractTestEnvironment;
import org.scijava.ops.function.Computers;
import org.scijava.ops.function.Inplaces;
import org.scijava.ops.function.Producer;

/**
 * Tests {@link OpBuilder}.
 * 
 * For each arity, we test the following matches and run the following commands 
 * based on the information scenarios.
 * 
 * * Input TYPES are given (IT) 
 * 	1) The output is unspecified (OU): 
 * 		a) match: Function, Inplace
 * 		b) run: none
 * 	2) The output type is given (OT): 
 * 		a) match: Function, Computer
 * 		b) run: none
 *  
 * * Input VALUES are given (IV) (N.B. this case applies for Arity0):
 * 	1) The output is unspecified (OU): 
 * 		a) match: Function, Inplace
 * 		b) run: apply, mutate
 * 	2) The output type is given (OT): 
 * 		a) match: Function, Computer
 * 		b) run: apply
 * 	3) The output value is given (OV): 
 * 		a) match: Computer
 *  	b) run: compute
 * 
 * @author Curtis Rueden
 */
public class OpBuilderTest extends AbstractTestEnvironment {

	final double[] halves = new double[10];
	{
		for (int i = 0; i < halves.length; i++)
			halves[i] = i / 2.;
	}

	// -- 0-ary --

	/** Matches a nullary function in a vacuum. */
	@Test
	public void testArity0_OU_matchF() {
		final Producer<?> op = name("test.addDoubles").input().producer();
		final Object result = op.create();
		assertEquals(0., result);
	}

	/** Runs a nullary function in a vacuum. */
	@Test
	public void testArity0_OU_runF() {
		final Object result = name("test.addDoubles").input().create();
		assertEquals(0., result);
	}

	/** Matches a nullary function with a given output type. */
	@Test
	public void testArity0_OT_matchF() {
		final Double expected = 0.;
		final Producer<Double> op = //
				name("test.addDoubles").input().outType(Double.class).producer();
		final Object result = op.create();
		assertEquals(result, expected);
	}

	/** Matches a nullary computer with a given output type. */
	@Test
	public void testArity0_OT_matchC() {
		final double[] actual = { 1, 2, 3 };
		final double[] expected = { 0, 0, 0 };
		final Computers.Arity0<double[]> op = //
				name("test.addArrays").input().outType(double[].class).computer();
		op.compute(actual);
		assertTrue(Arrays.equals(actual, expected));
	}

	/** Runs a nullary function with a given output type. */
	@Test
	public void testArity0_OT_runF() {
		final Double result = name("test.addDoubles").input().outType(Double.class).create();
		assert (0. == result);
	}

	/** Matches a nullary computer with a given output value. */
	@Test
	public void testArity0_OV_matchC() {
		double[] result = { 1, 2, 3 };
		double[] expected = { 0, 0, 0 };
		final Computers.Arity0<double[]> op = //
				name("test.addArrays").input().output(result).computer();
		op.compute(result);
		assertTrue(Arrays.equals(expected, result));
	}

	/** Runs a nullary computer with a given output value. */
	@Test
	public void testArity0_OV_runC() {
		double[] result = { 1, 2, 3 };
		double[] expected = { 0, 0, 0 };
		name("test.addArrays").input().output(result).compute();
		assertTrue(Arrays.equals(expected, result));
	}

	// -- 1-ary --

	/** Matches a unary function using input types only. */
	@Test
	public void testArity1_IT_OU_matchF() {
		final Double input = 1.;
		final Double expected = 1.;
		final Function<Double, ?> op = //
				name("test.addDoubles").inType(Double.class).function();
		final Object actual = op.apply(input);
		assertEquals(actual, expected);
	}

	/** Matches an inplace with a given output type. */
	@Test
	public void testArity1_IT_OU_matchI() {
		final double[] actual = { 1, 2, 3 };
		final double[] expected = { 1, 2, 3 };
		final Inplaces.Arity1<double[]> op = //
				name("test.mulArrays").inType(double[].class).inplace();
		op.mutate(actual);
		assertTrue(Arrays.equals(actual, expected));
	}

	/** Matches a unary function using input types + output type. */
	@Test
	public void testArity1_IT_OT_matchF() {
		final Double input = 1.;
		final Double expected = 1.;
		final Function<Double, Double> op = //
				name("test.addDoubles").inType(Double.class).outType(Double.class).function();
		final double actual = op.apply(input);
		assertEquals(actual, expected, 0.);
	}

	/** Matches a unary computer using input types + output type. */
	@Test
	public void testArity1_IT_OT_matchC() {
		double[] input = { 1, 2, 3 };
		double[] actual = { 0, 0, 0 };
		double[] expected = { 1, 2, 3 };
		final Computers.Arity1<double[], double[]> castToInt = //
				name("test.addArrays").inType(double[].class).outType(double[].class).computer();
		castToInt.compute(input, actual);
		assertTrue(Arrays.equals(actual, expected));
	}

	/** Matches a unary function using input value only. */
	@Test
	public void testArity1_IV_OU_matchF() {
		final Double input = 1.;
		final Double expected = 1.;
		final Function<Double, ?> op = //
				name("test.addDoubles").input(input).function();
		final Object actual = op.apply(input);
		assertEquals(actual, expected);
	}

	/** Matches a nullary inplace with a given output value. */
	@Test
	public void testArity1_IV_OU_matchI() {
		double[] result = { 1, 2, 3 };
		double[] expected = { 1, 2, 3 };
		final Inplaces.Arity1<double[]> op = //
				name("test.mulArrays").input(result).inplace();
		op.mutate(result);
		assertTrue(Arrays.equals(expected, result));
	}

	/** Runs a unary function using input value only. */
	@Test
	public void testArity1_IV_OU_runF() {
		final Double input = 1.;
		final Double expected = 1.;
		final Object actual = name("test.addDoubles").input(input).apply();
		assertEquals(actual, expected);
	}

	/** Runs a inplace with a given output value. */
	@Test
	public void testArity1_IV_OU_runI() {
		final double[] actual = { 1, 2, 3 };
		final double[] expected = { 1, 2, 3 };
		name("test.mulArrays").input(actual).mutate();
		assertTrue(Arrays.equals(actual, expected));
	}

	/** Matches a unary function using input value + output type. */
	@Test
	public void testArity1_IV_OT_matchF() {
		final Double input = 1.;
		final Double expected = 1.;
		final Function<Double, Double> op = //
				name("test.addDoubles").input(input).outType(Double.class).function();
		final double actual = op.apply(input);
		assertEquals(actual, expected, 0.);
	}

	/** Matches a unary computer using input types + output type. */
	@Test
	public void testArity1_IV_OT_matchC() {
		double[] input = { 1, 2, 3 };
		double[] actual = { 0, 0, 0 };
		double[] expected = { 1, 2, 3 };
		final Computers.Arity1<double[], double[]> castToInt = //
				name("test.addArrays").input(input).output(actual).computer();
		castToInt.compute(input, actual);
		assertTrue(Arrays.equals(actual, expected));
	}

	/** Runs a unary function using input value + output type. */
	@Test
	public void testArity1_IV_OT_runF() {
		final Double input = 1.;
		final Double expected = 1.;
		final Object actual = name("test.addDoubles").input(input).outType(Double.class).apply();
		assertEquals(actual, expected);
	}

	/** Matches a unary computer using input value + output value. */
	@Test
	public void testArity1_IV_OV_matchC() {
		double[] input = { 1, 2, 3 };
		double[] actual = { 0, 0, 0 };
		double[] expected = { 1, 2, 3 };
		final Computers.Arity1<double[], double[]> op = //
				name("test.addArrays").input(input).output(actual).computer();
		op.compute(input, actual);
		assertTrue(Arrays.equals(actual, expected));
	}

	/** Runs a unary computer using input value + output value. */
	@Test
	public void testArity1_IV_OV_runC() {
		double[] input = { 1, 2, 3 };
		double[] actual = { 0, 0, 0 };
		double[] expected = { 1, 2, 3 };
		name("test.addArrays").input(input).output(actual).compute();
		assertTrue(Arrays.equals(actual, expected));
	}

	// -- 2-ary --

	/** Matches a binary function using input types + output type. */
	@Test
	public void testArity2_IT_OU_matchF() {
		final Double input1 = 1.;
		final Double input2 = 1.;
		final Double expected = 2.;
		final BiFunction<Double, Double, ?> op = //
				name("test.addDoubles").inType(Double.class, Double.class).function();
		final Object actual = op.apply(input1, input2);
		assertEquals(actual, expected);
	}

	/** Matches a binary inplace1 using input types + output type. */
	@Test
	public void testArity2_IT_OU_matchI1() {
		final double[] actual1 = { 1, 2, 3 };
		final double[] actual2 = { 1, 2, 3 };
		final double[] expected = { 1, 4, 9 };
		final Inplaces.Arity2_1<double[], double[]> op = //
				name("test.mulArrays").inType(double[].class, double[].class).inplace1();
		op.mutate(actual1, actual2);
		assertTrue(Arrays.equals(actual1, expected));
		assertFalse(Arrays.equals(actual2, expected));
	}

	/** Matches a binary inplace1 using input types + output type. */
	@Test
	public void testArity2_IT_OU_matchI2() {
		final double[] actual1 = { 1, 2, 3 };
		final double[] actual2 = { 1, 2, 3 };
		final double[] expected = { 1, 4, 9 };
		final Inplaces.Arity2_2<double[], double[]> op = //
				name("test.mulArrays").inType(double[].class, double[].class).inplace2();
		op.mutate(actual1, actual2);
		assertTrue(Arrays.equals(actual2, expected));
		assertFalse(Arrays.equals(actual1, expected));
	}

	/** Matches a binary function using input types + output type. */
	@Test
	public void testArity2_IT_OT_matchF() {
		final Double input1 = 1.;
		final Double input2 = 1.;
		final Double expected = 2.;
		final BiFunction<Double, Double, Double> op = //
				name("test.addDoubles").inType(Double.class, Double.class).outType(Double.class).function();
		final Double actual = op.apply(input1, input2);
		assertEquals(actual, expected);
	}

	/** Matches a binary computer using input types + output type. */
	@Test
	public void testArity2_IT_OT_matchC() {
		double[] input1 = { 1, 2, 3 };
		double[] input2 = { 1, 2, 3 };
		double[] actual = { 0, 0, 0 };
		double[] expected = { 2, 4, 6 };
		final Computers.Arity2<double[], double[], double[]> op = //
				name("test.addArrays").inType(double[].class, double[].class).outType(double[].class).computer();
		op.compute(input1, input2, actual);
		assertTrue(Arrays.equals(actual, expected));
	}

	/** Matches a binary function using input values only. */
	@Test
	public void testArity2_IV_OU_matchF() {
		final Double input1 = 1.;
		final Double input2 = 1.;
		final Double expected = 2.;
		final BiFunction<Double, Double, ?> op = //
				name("test.addDoubles").input(input1, input2).function();
		final Object actual = op.apply(input1, input2);
		assertEquals(actual, expected);
	}

	/** Matches a binary inplace1 using input types + output type. */
	@Test
	public void testArity2_IV_OU_matchI1() {
		final double[] actual1 = { 1, 2, 3 };
		final double[] actual2 = { 1, 2, 3 };
		final double[] expected = { 1, 4, 9 };
		final Inplaces.Arity2_1<double[], double[]> op = //
				name("test.mulArrays").input(actual1, actual2).inplace1();
		op.mutate(actual1, actual2);
		assertTrue(Arrays.equals(actual1, expected));
		assertFalse(Arrays.equals(actual2, expected));
	}

	/** Matches a binary inplace1 using input types + output type. */
	@Test
	public void testArity2_IV_OU_matchI2() {
		final double[] actual1 = { 1, 2, 3 };
		final double[] actual2 = { 1, 2, 3 };
		final double[] expected = { 1, 4, 9 };
		final Inplaces.Arity2_2<double[], double[]> op = //
				name("test.mulArrays").input(actual1, actual2).inplace2();
		op.mutate(actual1, actual2);
		assertTrue(Arrays.equals(actual2, expected));
		assertFalse(Arrays.equals(actual1, expected));
	}

	/** Runs a binary function using input values only. */
	@Test
	public void testArity2_IV_OU_runF() {
		final Double input1 = 1.;
		final Double input2 = 1.;
		final Double expected = 2.;
		final Object actual = name("test.addDoubles").input(input1, input2).apply();
		assertEquals(actual, expected);
	}

	/** Matches a binary inplace1 using input types + output type. */
	@Test
	public void testArity2_IV_OU_runI1() {
		final double[] actual1 = { 1, 2, 3 };
		final double[] actual2 = { 1, 2, 3 };
		final double[] expected = { 1, 4, 9 };
		name("test.mulArrays").input(actual1, actual2).mutate1();
		assertTrue(Arrays.equals(actual1, expected));
		assertFalse(Arrays.equals(actual2, expected));
	}

	/** Matches a binary inplace1 using input types + output type. */
	@Test
	public void testArity2_IV_OU_runI2() {
		final double[] actual1 = { 1, 2, 3 };
		final double[] actual2 = { 1, 2, 3 };
		final double[] expected = { 1, 4, 9 };
		name("test.mulArrays").input(actual1, actual2).mutate2();
		assertTrue(Arrays.equals(actual2, expected));
		assertFalse(Arrays.equals(actual1, expected));
	}

	/** Matches a binary function using input value + output type. */
	@Test
	public void testArity2_IV_OT_matchF() {
		final Double input1 = 1.;
		final Double input2 = 1.;
		final Double expected = 2.;
		final BiFunction<Double, Double, Double> op = //
				name("test.addDoubles").input(input1, input2).outType(Double.class).function();
		final Double actual = op.apply(input1, input2);
		assertEquals(actual, expected);
	}

	/** Matches a binary computer using input value + output type. */
	@Test
	public void testArity2_IV_OT_matchC() {
		double[] input1 = { 1, 2, 3 };
		double[] input2 = { 1, 2, 3 };
		double[] actual = { 0, 0, 0 };
		double[] expected = { 2, 4, 6 };
		final Computers.Arity2<double[], double[], double[]> op = //
				name("test.addArrays").input(input1, input2).outType(double[].class).computer();
		op.compute(input1, input2, actual);
		assertTrue(Arrays.equals(actual, expected));
	}

	/** Runs a binary function using input values + output type. */
	@Test
	public void testArity2_IV_OT_runF() {
		final Double input1 = 1.;
		final Double input2 = 1.;
		final Double expected = 2.;
		final Object actual = name("test.addDoubles").input(input1, input2).outType(Double.class).apply();
		assertEquals(actual, expected);
	}

	/** Matches a binary computer using input value + output value. */
	@Test
	public void testArity2_IV_OV_matchC() {
		double[] input1 = { 1, 2, 3 };
		double[] input2 = { 1, 2, 3 };
		double[] actual = { 0, 0, 0 };
		double[] expected = { 2, 4, 6 };
		final Computers.Arity2<double[], double[], double[]> op = //
				name("test.addArrays").input(input1, input2).output(actual).computer();
		op.compute(input1, input2, actual);
		assertTrue(Arrays.equals(actual, expected));
	}

	/** Runs a binary computer using input value + output value. */
	@Test
	public void testArity2_IV_OV_runC() {
		double[] input1 = { 1, 2, 3 };
		double[] input2 = { 1, 2, 3 };
		double[] actual = { 0, 0, 0 };
		double[] expected = { 2, 4, 6 };
		name("test.addArrays").input(input1, input2).output(actual).compute();
		assertTrue(Arrays.equals(actual, expected));
	}

	// -- Helper methods --

	private OpBuilder name(String opName) {
		return new OpBuilder(ops, opName);
	}
}

File: scijava/buildSrc/src/main/kotlin/opsEngine/test/OpBuilderTest.kt
package opsEngine.test

import Generator
import license
import dontEdit
import kotlin.math.pow

open class OpBuilderTest : Generator() {

    //.include templates/main/java/org/scijava/ops/engine/util/Globals.list
    //    arities = (0..maxArity).collect()

    val inputDoubles get() = (0..arity).map { "Double" }

    val genericDoubles get() = '<' + inputDoubles.joinToString() + '>'

    val args
        get() = when (arity) {
            0 -> ""
            1 -> "input"
            else -> (1..arity).joinToString { "input$it" }
        }

    val inputArrays get() = (0 until arity).map { "double[]" }

    val arrayGenerics get() = '<' + inputArrays.joinToString() + '>'

    val computerArrayGenerics: String
        get() {
            arity++
            return arrayGenerics.also { arity-- }
        }

    //[OpBuilderTest.java]
    val wildcardOutputDoubles: String
        get() {
            arity--
            return inputDoubles.joinToString(prefix = "<", postfix = ", ?>").also { arity++ }
        }

    val doubleClassString: String
        get() {
            arity--
            return inputDoubles.joinToString { "$it.class" }.also { arity++ }
        }

    val arrayClassString get() = inputArrays.joinToString { "$it.class" }

    fun functionInputName(num: Int) = if (arity == 1) "input" else "input$num"

    val inplaceExpectedArray get() = "{ " + (1..3).joinToString { "" + it.toFloat().pow(arity) } + " }"

    val computerExpectedArray get() = "{ " + (1..3).joinToString { "" + arity * it } + " }"

    fun inplaceActualArg(i: Int) = if (arity == 1) "input" else "input$i"

    fun inplaceMethod(i: Int) = if (arity == 1) "inplace" else "inplace$i"

    fun mutateMethod(i: Int) = if (arity == 1) "mutate" else "mutate$i"

    override fun generate() {
        +"""
$license

package org.scijava.ops.engine;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.Arrays;
import java.util.function.BiFunction;
import java.util.function.Function;

import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.scijava.function.Computers;
import org.scijava.function.Functions;
import org.scijava.function.Inplaces;
import org.scijava.function.Producer;
import org.scijava.ops.api.OpBuilder;

/**
 * Tests {@link OpBuilder}. For each arity, we test the following matches and
 * run the following commands based on the information scenarios.
 * <p>
 * Input TYPES are given (IT):
 * <p>
 * <ol>
 * <li>The output is unspecified (OU):</li>
 * <ol type="a">
 * <li>match: Function, Inplace</li>
 * <li>run: none</li>
 * </ol>
 * <li>The output type is given (OT):</li>
 * <ol type="a">
 * <li>match: Function, Computer</li>
 * <li>run: none</li>
 * </ol>
 * </ol>
 * Input VALUES are given (IV) (N.B. this case applies for Arity0):
 * <p>
 * <ol>
 * <li>The output is unspecified (OU):</li>
 * <ol type="a">
 * <li>match: Function, Inplace</li>
 * <li>run: apply, mutate</li>
 * </ol>
 * <li>The output type is given (OT):</li>
 * <ol type="a">
 * <li>match: Function, Computer</li>
 * <li>run: apply</li>
 * </ol>
 * <li>The output value is given (OV):</li>
 * <ol type="a">
 * <li>match: Computer</li>
 * <li>run: compute</li>
 * </ol>
 * </ol>
 * 
 * @author Curtis Rueden
 * @author Gabriel Selzer
 */
public class OpBuilderTest extends AbstractTestEnvironment {

	@BeforeAll
	public static void AddNeededOps() {
		ops.register(new OpBuilderTestOps());
	}

	final double[] halves = new double[10];
	{
		for (int i = 0; i < halves.length; i++)
			halves[i] = i / 2.;
	}

	// -- 0-ary --

	/** Matches a nullary function in a vacuum. */
	@Test
	public void testArity0_OU_matchF() {
		final Producer<?> op = name("test.addDoubles").input().producer();
		final Object result = op.create();
		assertEquals(0., result);
	}

	/** Runs a nullary function in a vacuum. */
	@Test
	public void testArity0_OU_runF() {
		final Object result = name("test.addDoubles").input().create();
		assertEquals(0., result);
	}

	/** Matches a nullary function with a given output type. */
	@Test
	public void testArity0_OT_matchF() {
		final Double expected = 0.;
		final Producer<Double> op = //
			name("test.addDoubles").input().outType(Double.class).producer();
		final Object result = op.create();
		assertEquals(result, expected);
	}

	/** Matches a nullary computer with a given output type. */
	@Test
	public void testArity0_OT_matchC() {
		final double[] actual = { 1, 2, 3 };
		final double[] expected = { 0, 0, 0 };
		final Computers.Arity0<double[]> op = //
			name("test.addArrays").input().outType(double[].class).computer();
		op.compute(actual);
		assertTrue(Arrays.equals(actual, expected));
	}

	/** Runs a nullary function with a given output type. */
	@Test
	public void testArity0_OT_runF() {
		final Double result = name("test.addDoubles").input().outType(Double.class)
			.create();
		assert (0. == result);
	}

	/** Matches a nullary computer with a given output value. */
	@Test
	public void testArity0_OV_matchC() {
		double[] result = { 1, 2, 3 };
		double[] expected = { 0, 0, 0 };
		final Computers.Arity0<double[]> op = //
			name("test.addArrays").input().output(result).computer();
		op.compute(result);
		assertTrue(Arrays.equals(expected, result));
	}

	/** Runs a nullary computer with a given output value. */
	@Test
	public void testArity0_OV_runC() {
		double[] result = { 1, 2, 3 };
		double[] expected = { 0, 0, 0 };
		name("test.addArrays").input().output(result).compute();
		assertTrue(Arrays.equals(expected, result));
	}"""
        forEachArity(1..maxArity) {
            +"""

	// -- $arity-ary --

	/** Matches a $arity-arity function using input types only. */
	@Test
	public void testArity${arity}_IT_OU_matchF() {"""
            for (a in 1..arity)
                +"""
		final Double ${functionInputName(a)} = 1.;"""

            +"""
		final Double expected = $arity.;
		final $functionArity$wildcardOutputDoubles op = //
			name("test.addDoubles") //
				.inType($doubleClassString) //
				.function();
		final Object actual = op.apply($args);
		assertEquals(actual, expected);
	}
"""
            for (i in 1..arity) {
                +"""
	/** Matches a $arity-arity inplace$i with a given output type. */
	@Test
	public void testArity${arity}_IT_OU_matchI$i() {"""
                for (a in 1..arity)
                    +"""
		final double[] ${functionInputName(a)} = { 1, 2, 3 };"""

                +"""
		final double[] expected = $inplaceExpectedArray;
		final ${inplaceType(i)}$arrayGenerics op = //
			name("test.mulArrays${arity}_$i") //
				.inType($arrayClassString) //
				.${inplaceMethod(i)}();
		op.mutate($args);
		assertTrue(Arrays.equals(${inplaceActualArg(i)}, expected));
	}
"""
            }
            +"""
	/** Matches a $arity-arity function using input types + output type. */
	@Test
	public void testArity${arity}_IT_OT_matchF() {"""
            for (a in 1..arity)
                +"""
		final Double ${functionInputName(a)} = 1.;"""

            +"""
		final Double expected = $arity.;
		final $functionArity$genericDoubles op = //
			name("test.addDoubles").inType($doubleClassString) //
				.outType(Double.class).function();
		final double actual = op.apply($args);
		assertEquals(actual, expected, 0.);
	}

	/** Matches a $arity-arity computer using input types + output type. */
	@Test
	public void testArity${arity}_IT_OT_matchC() {"""
            for (a in 1..arity)
                +"""
		final double[] ${functionInputName(a)} = { 1, 2, 3 };"""

            +"""
		double[] actual = { 0, 0, 0 };
		double[] expected = $computerExpectedArray;
		final $computerArity$computerArrayGenerics op = //
			name("test.addArrays") //
				.inType($arrayClassString) //
				.outType(double[].class).computer();
		op.compute($args, actual);
		assertTrue(Arrays.equals(actual, expected));
	}

	/** Matches a $arity-arity function using input value only. */
	@Test
	public void testArity${arity}_IV_OU_matchF() {"""
            for (a in 1..arity)
                +"""
		final Double ${functionInputName(a)} = 1.;"""

            +"""
		final Double expected = $arity.;
		final $functionArity$wildcardOutputDoubles op = //
			name("test.addDoubles") //
				.input($args) //
				.function();
		final Object actual = op.apply($args);
		assertEquals(actual, expected);
	}
"""
            for (i in 1..arity) {
                +"""
	/** Matches a $arity-arity inplace$i with a given output value. */
	@Test
	public void testArity${arity}_IV_OU_matchI$i() {"""
                for (a in 1..arity)
                    +"""
		final double[] ${functionInputName(a)} = { 1, 2, 3 };"""

                +"""
		double[] expected = $inplaceExpectedArray;
		final ${inplaceType(i)}$arrayGenerics op = //
			name("test.mulArrays${arity}_$i").input($args) //
				.${inplaceMethod(i)}();
		op.mutate($args);
		assertTrue(Arrays.equals(expected, ${functionInputName(i)}));
	}
"""
            }
            +"""
	/** Runs a $arity-arity function using input value only. */
	@Test
	public void testArity${arity}_IV_OU_runF() {"""
            for (a in 1..arity)
                +"""
		final Double ${functionInputName(a)} = 1.;"""

            +"""
		final Double expected = $arity.;
		final Object actual = name("test.addDoubles") //
			.input($args) //
			.apply();
		assertEquals(actual, expected);
	}
"""
            for (i in 1..arity) {
                +"""
	/** Runs a $arity-arity inplace${i} with a given output value. */
	@Test
	public void testArity${arity}_IV_OU_runI$i() {"""
                for (a in 1..arity)
                    +"""
		final double[] ${functionInputName(a)} = { 1, 2, 3 };"""

                +"""
		final double[] expected = $inplaceExpectedArray;
		name("test.mulArrays${arity}_$i") //
			.input($args) //
			.${mutateMethod(i)}();
		assertTrue(Arrays.equals(${functionInputName(i)}, expected));
	}
"""
            }
            +"""
	/** Matches a $arity-arity function using input value + output type. */
	@Test
	public void testArity${arity}_IV_OT_matchF() {"""
            for (a in 1..arity)
                +"""
		final Double ${functionInputName(a)} = 1.;"""

            +"""
		final Double expected = $arity.;
		final $functionArity$genericDoubles op = //
			name("test.addDoubles").input($args) //
				.outType(Double.class).function();
		final double actual = op.apply($args);
		assertEquals(actual, expected, 0.);
	}

	/** Matches a unary computer using input types + output type. */
	@Test
	public void testArity${arity}_IV_OT_matchC() {"""
            for (a in 1..arity)
                +"""
		double[] ${functionInputName(a)} = { 1, 2, 3 };"""

            +"""
		double[] actual = { 0, 0, 0 };
		double[] expected = $computerExpectedArray;
		final $computerArity$computerArrayGenerics op = //
			name("test.addArrays") //
				.input($args) //
				.output(actual).computer();
		op.compute($args, actual);
		assertTrue(Arrays.equals(actual, expected));
	}

	/** Runs a unary function using input value + output type. */
	@Test
	public void testArity${arity}_IV_OT_runF() {"""
            for (a in 1..arity)
                +"""
		final Double ${functionInputName(a)} = 1.;"""

            +"""
		final Double expected = $arity.;
		final Object actual = name("test.addDoubles") //
			.input($args) //
			.outType(Double.class).apply();
		assertEquals(actual, expected);
	}

	/** Matches a unary computer using input value + output value. */
	@Test
	public void testArity${arity}_IV_OV_matchC() {"""
            for (a in 1..arity)
                +"""
		double[] ${functionInputName(a)} = { 1, 2, 3 };"""

            +"""
		double[] actual = { 0, 0, 0 };
		double[] expected = $computerExpectedArray;
		final $computerArity$computerArrayGenerics op = //
			name("test.addArrays") //
				.input($args) //
				.output(actual).computer();
		op.compute($args, actual);
		assertTrue(Arrays.equals(actual, expected));
	}

	/** Runs a unary computer using input value + output value. */
	@Test
	public void testArity${arity}_IV_OV_runC() {"""
            for (a in 1..arity)
                +"""
		double[] ${functionInputName(a)} = { 1, 2, 3 };"""

            +"""
		double[] actual = { 0, 0, 0 };
		double[] expected = $computerExpectedArray;
		name("test.addArrays") //
			.input($args) //
			.output(actual).compute();
		assertTrue(Arrays.equals(actual, expected));
	}"""
        }
        +"""

	// -- Helper methods --

	private OpBuilder name(String opName) {
		return ops.op(opName);
	}
}
"""
    }
}